wallet/chainntfns.go

```diff
@@ -329,3 +341,152 @@ func (w *Wallet) addRelevantTx(dbtx walletdb.ReadWriteTx, rec *wtxmgr.TxRecord,
+// to ensure we do not miss any relevant events throughout rescans.
+func (w *Wallet) birthdaySanityCheck() error {
+	// We'll start by acquiring our chain backend client as we'll be
+	// querying it for blocks.
+	chainClient, err := w.requireChainClient()
+	if err != nil {
+		return err
+	}
+
+	// We'll then fetch our wallet's birthday timestamp and block.
+	birthdayTimestamp := w.Manager.Birthday()
+	var birthdayBlock waddrmgr.BlockStamp
+	err = walletdb.View(w.db, func(tx walletdb.ReadTx) error {
+		var err error
+		ns := tx.ReadBucket(waddrmgrNamespaceKey)
+		birthdayBlock, err = w.Manager.BirthdayBlock(ns)
+		return err
+	})
+
+	switch {
+	// If our wallet's birthday block has not been set yet, then this is our
+	// initial sync, so we'll defer setting it until then.
+	case waddrmgr.IsError(err, waddrmgr.ErrBirthdayBlockNotSet):
+		return nil
+
+	// Otherwise, we'll return the error if there was one.
+	case err != nil:
+		return err
+	}
+
+	log.Debugf("Starting sanity check for the wallet's birthday block "+
+		"from: height=%d, hash=%v", birthdayBlock.Height,
+		birthdayBlock.Hash)
+
+	// Now, we'll need to determine if our block correctly reflects our
+	// timestamp. To do so, we'll fetch the block header and check its
+	// timestamp.
+	candidate := birthdayBlock
+	header, err := chainClient.GetBlockHeader(&candidate.Hash)
+	if err != nil {
+		return fmt.Errorf("unable to get header for block hash %v: %v",
+			candidate.Hash, err)
+	}
+
+	// We'll go back a day worth of blocks in the chain until we find a
+	// block whose timestamp is below our birthday timestamp.
+	heightDelta := int32(144)
+	for birthdayTimestamp.Before(header.Timestamp) {
+		// If the birthday block has reached genesis, then we can exit
+		// our search as there exists no data before this point.
+		if candidate.Height == 0 {
+			break
+		}
+
+		candidate.Height -= heightDelta
+		if candidate.Height < 0 {
+			candidate.Height = 0
+		}
+		hash, err := chainClient.GetBlockHash(int64(candidate.Height))
+		if err != nil {
+			return fmt.Errorf("unable to get block hash for "+
+				"height %d: %v", candidate.Height, err)
+		}
+		candidate.Hash = *hash
+
+		header, err = chainClient.GetBlockHeader(hash)
+		if err != nil {
+			return fmt.Errorf("unable to get header for block "+
+				"hash %v: %v", candidate.Hash, err)
+		}
+
+		log.Debugf("Checking next birthday block candidate: "+
+			"height=%d, hash=%v, timestamp=%v",
+			candidate.Height, candidate.Hash,
+			header.Timestamp)
+	}
+
+	// To ensure we have a reasonable birthday block, we'll make sure it
+	// respects our birthday timestamp and it is within a reasonable delta.
+	// The birthday has already been adjusted to two days in the past of the
+	// actual birthday, so we'll make our expected delta to be within two
+	// hours of it to account for the network-adjusted time and prevent
+	// fetching more unnecessary blocks.
+	timestampDelta := birthdayTimestamp.Sub(header.Timestamp)
+	for timestampDelta > 2*time.Hour {
```

Comment:
I don't think this loop is worth it. At most we can gain 144 blocks, which doesn't seem too much extra to scan.

Reply:
It is needed due to the height estimate of the migration potentially being wrong. On mainnet this isn't such an issue, but on testnet, like you pointed out, the estimate is much farther back than it should be, so the sanity check should be able to detect this and look for a more reasonable value.

Reply:
As the goal is to minimize the rescan duration (since users _really_ hated the migration time from 0.5) logic that attempts to bridge the gap is warranted IMO. I think with real world tests, we'll be able to get a feel for how off the estimates are on mainnet/testnet. At that point, we can make a decisions w.r.t the fate of this fragment.

Reply:
Correct. On mainnet, the height estimate was off by about 30K blocks in the past, which is still a pretty big improve over the genesis block IMO. On testnet, it was ~1 million.

Reply:
Ah, I see. I was assuming the candidate started out `>` birthday before the first loop.